File: src/main/java/io/github/ajoz/workshop/fp/java/tools/ValueIterator.java
package io.github.ajoz.workshop.fp.java.tools;

import java.util.Iterator;
import java.util.NoSuchElementException;

public final class ValueIterator<T> implements Iterator<T> {
    private final T[] mValues;
    private int mCurrentIndex;

    @SafeVarargs
    ValueIterator(final T... values) {
        this.mValues = values;
    }

    @Override
    public boolean hasNext() {
        return mCurrentIndex < mValues.length;
    }

    @Override
    public T next() {
        if (mCurrentIndex >= mValues.length) {
            throw new NoSuchElementException("This iterator has only: " + mValues.length + " values!");
        }

        final T value = mValues[mCurrentIndex];
        mCurrentIndex++;
        return value;
    }

    @Override
    public void remove() {
        throw new UnsupportedOperationException();
    }
}

File: src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/tools/ValueIterator.kt
package io.github.ajoz.workshop.fp.kotlin.tools

import java.util.NoSuchElementException

class ValueIterator<A>(private vararg val values: A) : Iterator<A> {
    private var currentIndex: Int = 0

    override fun hasNext() = currentIndex < values.size

    override fun next(): A {
        if (currentIndex >= values.size) {
            throw NoSuchElementException("This iterator has only: " + values.size + " values!")
        }

        return values[currentIndex++]
    }
}